opentrons_drivers/common/actions.py

```diff
@@ -1,37 +1,102 @@
 import time
-import os
-from inspect import signature
-import sys
+from typing import cast
+from typing import Dict, Callable, List, TypeVar
+from opentrons_drivers.common.custom_types import ActionFn
+from opentrons_drivers.common.custom_types import StockWell, CoreWell, StaticCtx, JSONType
+from opentrons.protocol_api.labware import Well
+from opentrons.protocol_api.instrument_context import InstrumentContext
 
-def call_with_needed(fn, **kwargs):
-    needed = signature(fn).parameters
-    return fn(**{k: v for k, v in kwargs.items() if k in needed})
+#---------- Registries of possible functions ----------
+
+ACTION_REGISTRY: Dict[str, ActionFn] = {} # is exported
+LIQUID_METHODS: Dict[str, Callable[..., object]] = {} # type: ignore[misc]
+
+F = TypeVar("F", bound=Callable[..., object]) # type: ignore[misc]
+
+# Shared registry decorator factory
+def make_registry_decorator(registry: Dict[str, F]) -> Callable[[str], Callable[[F], F]]: # type: ignore[misc]
+    def register(name: str) -> Callable[[F], F]: # type: ignore[misc]
+        def decorator(fn: F) -> F: # type: ignore[misc]
+            registry[name] = fn
+            return fn
+        return decorator
+    return register
```

Comment:
I think this pattern will work well. Let's make it a bit more obvious what it's doing and improve typing a bit. 

`from collections.abc import Callable`.

 ```suggestion
ACTION_REGISTRY: dict[str, ActionFn] = {}  # is exported
LIQUID_METHODS: dict[str, Callable[..., object]] = {}

F = TypeVar("F", bound=Callable[..., object])

def make_registry_decorator[T: Callable[..., object]](
    registry: dict[str, T]
) -> Callable[[str], Callable[[T], T]]:
    """Create a decorator factory for registering functions in a registry.
    
    Args:
        registry: Dictionary to store registered functions
        
    Returns:
        A decorator factory that creates registration decorators
    """
    def register(name: str) -> Callable[[T], T]:
        """Create a decorator that registers a function with the given name.
        
        Args:
            name: Name to register the function under
            
        Returns:
            Decorator that registers the function and returns it unchanged
        """
        def decorator(fn: T) -> T:
            registry[name] = fn
            return fn
        return decorator
    return register
```